src/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs

```diff
@@ -2,291 +2,78 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System.Buffers.Binary;
-using System.Diagnostics;
+using System.Net.Quic.Implementations;
 using System.Net.Security;
-using System.Net.Sockets;
 using System.Threading;
 using System.Threading.Tasks;
 
 namespace System.Net.Quic
 {
     public sealed class QuicConnection : IDisposable
     {
-        private readonly bool _isClient;
-        private bool _disposed = false;
-        private IPEndPoint _remoteEndPoint;
-        private IPEndPoint _localEndPoint;
-        private object _syncObject = new object();
-
-        // !!! TEMPORARY FOR QUIC MOCK SUPPORT
-        private readonly bool _mock = false;
-        private Socket _socket = null;
-        private IPEndPoint _peerListenEndPoint = null;
-        private TcpListener _inboundListener = null;
-        private long _nextOutboundBidirectionalStream;
-        private long _nextOutboundUnidirectionalStream;
+        private readonly QuicConnectionProvider _provider;
 
         /// <summary>
         /// Create an outbound QUIC connection.
         /// </summary>
         /// <param name="remoteEndPoint">The remote endpoint to connect to.</param>
         /// <param name="sslClientAuthenticationOptions">TLS options</param>
         /// <param name="localEndPoint">The local endpoint to connect from.</param>
-        /// <param name="mock">Use mock QUIC implementation.</param>
-        // !!! TEMPORARY FOR QUIC MOCK SUPPORT: Remove "mock" parameter before shipping
-        public QuicConnection(IPEndPoint remoteEndPoint, SslClientAuthenticationOptions sslClientAuthenticationOptions, IPEndPoint localEndPoint = null, bool mock = false)
+        public QuicConnection(IPEndPoint remoteEndPoint, SslClientAuthenticationOptions sslClientAuthenticationOptions, IPEndPoint localEndPoint = null)
+            : this(remoteEndPoint, sslClientAuthenticationOptions, localEndPoint, implementationProvider: null)
```

Comment:
Shouldn't this provide a default value for implementationProvider? The code as is looks like it will null ref.